### PR TITLE
Add EdgeConnect CRD

### PR DESCRIPTION
**Pull request by aorcholski**
## Description

Add `EdgeConnect` CustomResourceDefinition. Both `DynaKube` and `EdgeConnect` CRDs are created/applied using existing commands (`make manifests`, `make deploy/...`). `Common/crd/dynatrace-operator-crd.yaml` helm template contains both CRDs so the `installCRD` field controls installation of both CRDs.

`kubectl explain` command can be used to check what fields exist and what they do.

## How can this be tested?
```
# make deploy/kubernetes/no-csi
# kubectl explain edgeconnects.dynatrace.com.spec
# kubectl explain edgeconnects.dynatrace.com.status
```
Default values (`autoUpdate`, `replicas`):
```
# kubectl -n dynatrace apply -f - <<EOF
apiVersion: dynatrace.com/v1alpha1
kind: EdgeConnect
metadata:
  name: edgeconnect
  namespace: dynatrace
spec:
  apiServer: "aa.bb.com"

  oauth:
    clientSecret: "secret"
    endpoint: "endpoint"
    resource: "urn"
EOF

# kubectl -n dynatrace get  edgeconnect.dynatrace.com/edgeconnect
NAME          APISERVER   STATUS   AGE
edgeconnect   aa.bb.com            15s

# kubectl -n dynatrace describe  edgeconnect.dynatrace.com/edgeconnect
Name:         edgeconnect
Namespace:    dynatrace
Labels:       <none>
Annotations:  <none>
API Version:  dynatrace.com/v1alpha1
Kind:         EdgeConnect
Metadata:
  Creation Timestamp:  2023-07-25T13:57:00Z
  Generation:          3
  Resource Version:    14898705
  UID:                 ec0dc915-7451-4c02-8dc1-0a4800546fd7
Spec:
  API Server:   aa.bb.com
  Auto Update:  true
  Oauth:
    Client Secret:  secret
    Endpoint:       endpoint
    Resource:       urn
  Replicas:         1
Events:             <none>

```
Basic validation - required fields:
 - `oauth.clientSecret`
 - `oauth.endpoint`
 -  `oauth.resource`
 -  `autoUpdate`
 - `replicas`
```
# kubectl -n dynatrace apply -f - <<EOF
apiVersion: dynatrace.com/v1alpha1
kind: EdgeConnect
metadata:
  name: edgeconnect
  namespace: dynatrace
spec:
  apiServer: "aa.bb.com"

  oauth:
                    
    endpoint: "endpoint"
    resource: "urn"
EOF

The EdgeConnect "edgeconnect" is invalid: spec.oauth.clientSecret: Required value

```

## Checklist

~~- [ ] Unit tests have been updated/added~~
- [x] PR is labeled accordingly with a single label
- [x] I have read and understood the [contribution guidelines](https://github.com/Dynatrace/dynatrace-operator/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+	// Location of the Dynatrace API to connect to, including your specific environment UUID
+	// +required
+	// +kubebuilder:validation:Required
+	ApiServer string `json:"apiServer"`
+
+	// EdgeConnect uses the OAuth client to authenticate itself with the Dynatrace platform.
+	// +required
+	// +kubebuilder:validation:Required
+	Oauth OAuthSpec `json:"oauth"`
```
Comment (StefanHauth) [inline src/api/v1alpha1/edgeconnect/edgeconnect_types.go:25]:
```suggestion
	OAuth OAuthSpec `json:"oauth"`
```

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+	// Optional: indicates version of the EdgeConnect image to use
+	// +kubebuilder:validation:Optional
+	Tag string `json:"tag,omitempty"`
+}
+
+// EdgeConnectStatus defines the observed state of EdgeConnect
+type EdgeConnectStatus struct {
+	// Defines the current state (Running, Updating, Error, ...)
+	Phase status.PhaseType `json:"phase,omitempty"`
```
Comment (StefanHauth) [inline src/api/v1alpha1/edgeconnect/edgeconnect_types.go:106]:
```suggestion
	DeploymentPhase status.DeploymentPhase `json:"phase,omitempty"`
```
```suggestion
	DeploymentState status.DeploymentState `json:"phase,omitempty"`
```

**Review by StefanHauth — CHANGES_REQUESTED** [review]

File: src/api/status/phase.go [hunk 3]
```diff
@@ -0,0 +1,11 @@
+// +kubebuilder:object:generate=true
+// +k8s:openapi-gen=true
+package status
+
+type PhaseType string
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply src/api/status/phase.go:5]: Fixed

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go (hunk 1 above)
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go:25]: fixed

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go (hunk 2 above)
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go:106]: fixed

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+)
+
+// EdgeConnectSpec defines the desired state of EdgeConnect
+type EdgeConnectSpec struct {
+	// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
+
+	// Location of the Dynatrace API to connect to, including your specific environment UUID
+	// +required
+	// +kubebuilder:validation:Required
+	ApiServer string `json:"apiServer"`
```
[the remark this answers is not in the document]
Comment (chrismuellner) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go:20]: this field won't be a [url](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/Web_mechanics/What_is_a_URL), with protocol/path/parameters, but just the hostname for the 3rd gen tenant.

[the remark this answers is not in the document]
Comment (StefanHauth) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go:20]: Understood. Then it's perfectly named.

**Review by StefanHauth — APPROVED** [review]
LGTM, but i guess a second opinion wouldn't hurt

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	// Optional: set topology spread constraints for the EdgeConnect pods
+	// +kubebuilder:validation:Optional
+	TopologySpreadConstraints []corev1.TopologySpreadConstraint `json:"topologySpreadConstraints,omitempty"`
+}
+
+type OAuthSpec struct {
+	// The secret from your OAuth client generation
+	// +required
+	// +kubebuilder:validation:Required
+	ClientSecret string `json:"clientSecret"`
```
Comment (chrismuellner) [inline src/api/v1alpha1/edgeconnect/edgeconnect_types.go:82]: slightly different wording in the [edgeconnect doc](https://docs.dynatrace.com/platform/shortlink/edgeconnect#config), please align that.

We also don't need the `The` at the start

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	// +kubebuilder:validation:Required
+	Endpoint string `json:"endpoint"`
+	// URN identifying your account. You get the URN when creating the OAuth client
+	// +required
+	// +kubebuilder:validation:Required
+	Resource string `json:"resource"`
+}
+
+type ImageRefSpec struct {
+	// Optional: if specified, indicates the EdgeConnect repository to use
```
Comment (chrismuellner) [inline src/api/v1alpha1/edgeconnect/edgeconnect_types.go:94]:
```suggestion
	// Custom EdgeConnect image repository
```

"Optional" and "if specified" is redundant.

Comment (chrismuellner) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go:94]: We could use the `+kubebuilder:example` marker to show what an expected value for this field (and some others where it makes sense) is, e.g. `docker.io/edgeconnect`

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+	// +kubebuilder:validation:Optional
+	Tag string `json:"tag,omitempty"`
+}
+
+// EdgeConnectStatus defines the observed state of EdgeConnect
+type EdgeConnectStatus struct {
+	// Defines the current state (Running, Updating, Error, ...)
+	DeploymentPhase status.DeploymentPhase `json:"phase,omitempty"`
+
+	// State of the current image
```
Comment (chrismuellner) [inline src/api/v1alpha1/edgeconnect/edgeconnect_types.go:108]:
```suggestion
	// Version used for the Edgeconnect image
```

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+	Version status.VersionStatus `json:"version,omitempty"`
+
+	// Indicates when the instance was last updated
```
Comment (chrismuellner) [inline src/api/v1alpha1/edgeconnect/edgeconnect_types.go:111]:
```suggestion
	// Indicates when the resource was last updated
```

File: src/api/v1alpha1/edgeconnect/edgeconnect_webhook.go [hunk 9]
```diff
@@ -0,0 +1 @@
+package edgeconnect
```
Comment (chrismuellner) [inline src/api/v1alpha1/edgeconnect/edgeconnect_webhook.go:1]: do we need this placeholder?

File: src/api/v1beta1/dynakube/test/certs_test.go [hunk 10]
```diff
@@ -4,7 +4,7 @@ import (
 	"context"
 	"testing"
```
Comment (chrismuellner) [inline src/api/v1beta1/dynakube/test/certs_test.go:6]: why are these tests in their own `test` folder?

**Review by chrismuellner — CHANGES_REQUESTED** [review]

File: src/api/v1alpha1/dynakube/dynakube_types.go [hunk 11]
```diff
@@ -1,12 +1,17 @@
...
 import (
+	"github.com/Dynatrace/dynatrace-operator/src/api/status"
+	"github.com/Dynatrace/dynatrace-operator/src/api/v1alpha1"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // DynaKubeSpec defines the desired state of DynaKube
-// +k8s:openapi-gen=true
+// +-  k8s:openapi-gen=true
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply src/api/v1alpha1/dynakube/dynakube_types.go:14]: it's unwanted change

File: config/crd/bases/dynatrace.com_edgeconnects.yaml [hunk 12]
```diff
@@ -0,0 +1,596 @@
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.12.0
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply config/crd/bases/dynatrace.com_edgeconnects.yaml:6]: fixed

File: src/api/v1alpha1/edgeconnect/edgeconnect_conversion.go [hunk 13]
```diff
@@ -0,0 +1 @@
+package edgeconnect
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_conversion.go:1]: removed

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EdgeConnectSpec defines the desired state of EdgeConnect
+type EdgeConnectSpec struct {
+	// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
+
+	// Location of the Dynatrace API to connect to, including your specific environment UUID
+	// +required
+	// +kubebuilder:validation:Required
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go:19]: +required not needed

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+
+	// Optional: restrict outgoing HTTP requests to your internal resources to specified hosts
+	// +kubebuilder:validation:Optional
+	HostRestrictions string `json:"hostRestrictions,omitempty"`
+
+	// Optional: image reference
+	// +kubebuilder:validation:Optional
+	ImageRef ImageRefSpec `json:"imageRef,omitempty"`
+
+	// Optional: enables automatic restarts of EdgeConnect pods in case a new version is available (the default value is: true)
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go:35]: removed

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+	// EdgeConnect uses the OAuth client to authenticate itself with the Dynatrace platform.
+	// +required
+	// +kubebuilder:validation:Required
+	OAuth OAuthSpec `json:"oauth"`
+
+	// Optional: restrict outgoing HTTP requests to your internal resources to specified hosts
+	// +kubebuilder:validation:Optional
+	HostRestrictions string `json:"hostRestrictions,omitempty"`
+
+	// Optional: image reference
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go:31]: fixed

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+	// +kubebuilder:validation:Required
+	ApiServer string `json:"apiServer"`
+
+	// EdgeConnect uses the OAuth client to authenticate itself with the Dynatrace platform.
+	// +required
+	// +kubebuilder:validation:Required
+	OAuth OAuthSpec `json:"oauth"`
+
+	// Optional: restrict outgoing HTTP requests to your internal resources to specified hosts
+	// +kubebuilder:validation:Optional
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go:28]: marker moved to the package level

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go (hunk 7 above)
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go:108]: fixed

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go (hunk 8 above)
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go:111]: fixed

File: src/api/v1alpha1/edgeconnect/edgeconnect_webhook.go (hunk 9 above)
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_webhook.go:1]: removed

File: src/api/v1beta1/dynakube/test/certs_test.go (hunk 10 above)
Comment (aorcholski) [reply src/api/v1beta1/dynakube/test/certs_test.go:6]: These test have been created in this way

File: config/crd/bases/kustomization.yaml [hunk 18]
```diff
@@ -0,0 +1,4 @@
+resources:
+- dynatrace.com_dynakubes.yaml
+- dynatrace.com_edgeconnects.yaml
+#+kubebuilder:scaffold:crdkustomizeresource
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply config/crd/bases/kustomization.yaml]: removed

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+	// Optional: enables automatic restarts of EdgeConnect pods in case a new version is available (the default value is: true)
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=true
+	AutoUpdate bool `json:"autoUpdate"`
+
+	// Optional: pull secret for your private registry
+	// +kubebuilder:validation:Optional
+	CustomPullSecret string `json:"customPullSecret,omitempty"`
+
+	// Optional: adds additional annotations for the EdgeConnect pods
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go]: fixed

File: src/api/v1beta1/dynakube/test/certs_test.go (hunk 10 above)
Comment (0sewa0) [reply src/api/v1beta1/dynakube/test/certs_test.go:6]: you can't move them, otherwise there will be a circular dependency ...

Comment (aorcholski) [reply src/api/v1beta1/dynakube/test/certs_test.go:6]:
```
package github.com/Dynatrace/dynatrace-operator/src/api/v1beta1/dynakube
        imports github.com/Dynatrace/dynatrace-operator/src/scheme
        imports github.com/Dynatrace/dynatrace-operator/src/scheme/fake: import cycle not allowed in test
```

File: src/api/v1alpha1/dynakube/dynakube_types.go [hunk 20]
```diff
@@ -226,10 +231,10 @@ type DynaKubeProxy struct {
 }
 
 // DynaKubeStatus defines the observed state of DynaKube
-// +k8s:openapi-gen=true
+// +- k8s:openapi-gen=true
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply src/api/v1alpha1/dynakube/dynakube_types.go]: fixed

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+	// +kubebuilder:validation:Required
+	Endpoint string `json:"endpoint"`
+	// URN identifying your account. You get the URN when creating the OAuth client
+	// +kubebuilder:validation:Required
+	Resource string `json:"resource"`
+}
+
+type ImageRefSpec struct {
+	// Custom EdgeConnect image repository
+	// +kubebuilder:example:="docker.io/edgeconnect"
```
[the remark this answers is not in the document]
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go]: fixed

File: src/api/v1alpha1/edgeconnect/edgeconnect_types.go (hunk 5 above)
Comment (aorcholski) [reply src/api/v1alpha1/edgeconnect/edgeconnect_types.go:69]: fixed

**Review by chrismuellner — APPROVED** [review]

**Pull request merged**